### DISCUSSION TITLE
Polygon Boolean Operations Library - iOverlay

### DISCUSSION THREAD

**Discussion by NailxSharipov** [The-OpenROAD-Project/OpenROAD#5864] [Show and tell]
Hi OpenRoad Community,

I’d like to introduce my library [iOverlay](https://github.com/iShape-Rust/iOverlay), which provides efficient 2D polygon Boolean operations. I believe this library could be particularly valuable for GDSII-related tasks and other polygon manipulations in EDA tools.

Over the past year, I’ve worked on implementing DRC rules and found that existing libraries often struggled with precision and performance, especially with large datasets. That’s why I developed iOverlay.

iOverlay has reached a stable version and is now ready for production use.
Here are some of its key features:

- **Precision**: Utilizes fixed-point arithmetic for high accuracy and reliable results, particularly useful for complex geometries.
- **Industry Standard Compatibility**: iOverlay operates using (i32, i32) for points, which aligns with GDSII standards.
- **Stability**: No risk of deadlocks or crashes, ensuring robustness.
- **Performance**: Optimized for handling large datasets and intricate shapes. I’ve also included a detailed [performance comparison](https://ishape-rust.github.io/iShape-js/overlay/performance/performance.html) that showcases its efficiency.
- **Documentation**: Comes with comprehensive explanations of how the library works and how to get started.

You can also read my [Article](https://dev.to/nail_sharipov_5d810d8cf71/2d-polygon-boolean-operations-4df7) for more information on how it works.

Looking forward to hearing your thoughts!

**Comment by maliberty** [comment]
You have a nice algorithm description.  We currently use Boost polygon for this purpose.  Do you have a comparison of performance and features?  Is there a C++ API?

**Reply by NailxSharipov** [reply]
I haven’t directly compared it with Boost Polygon yet, but I’m going to test it. Currently, there isn’t a C++ API, as I’ve primarily focused on Rust, considering the C++ ecosystem already has several mature options.

**Reply by maliberty** [reply]
C++ API would be needed for anything to happen here as that's our development language.  MT could be an advantage.

**Comment by maliberty** [comment]
A curiosity question: you use i32 for coordinates but how you handle intersection points that don't fall on an integer value when splitting segments?  For example the self-intersecting (0,0) (3,1) (0,3) (3,0) (0,0) ?

**Reply by NailxSharipov** [reply]
Yes, for intersection points, the library snaps them to the nearest integer grid corner. This isn't a truncation but a proper rounding to the nearest value, ensuring precision. To handle that, I implemented a custom 128-bit integer division to maintain accuracy.

The library offers three precision levels:

Absolute, Average and Auto.

![precision](https://github.com/user-attachments/assets/a2df6446-d6f6-423a-aa4a-ec691a9b013a)

Here I show 2 result:

1 with absolute presicion and 2 with Auto.

And of course nobody use this scale in practice. The library has f32/f64 api and it automatic rescale.

But it is still intresting to discuss.

**Reply by NailxSharipov** [reply]
The second good question will be what will happen when we change original geometry (by snaping it to a grid) and it generate a second level of intersection points.

**Comment by NailxSharipov** [comment]
@maliberty Finally, I found time to test the Boost library. Based on the results, it looks like Boost is an outsider in this race. I’m not a C++ expert, so there’s a small chance I may have done something wrong.

However, I think it’s better to attract some attention to this. It might make more sense to switch to **Clipper2**.

Here are the tests I conducted: [Performance Tests](https://ishape-rust.github.io/iShape-js/overlay/performance/performance.html)

**Comment by stefanottili** [comment]
You seem to use boost.geometry, please compare with boost.polygon.
If you have orthogonal or 45 degree data, take advantage of polygon90/45.
Recent technologies only allow orthogonal geometry.

**Comment by NailxSharipov** [comment]
Hi, I'm back :)

I finally found the time to build a dedicated library for orthogonal polygon operations: [xOverlay](https://github.com/iShape-Rust/xOverlay).

The results turned out quite well. In most benchmarks, **xOverlay** was **10–20× faster than Boost.Polygon**.

I found only one pathological case **Nested Squares** where **Boost.Polygon** performed significantly better.

I'd appreciate it if you guys @stefanottili, @maliberty  could take a look at the [benchmarks](https://ishape-rust.github.io/xOverlay/performance/) and share your thoughts.

**Reply by stefanottili** [reply]
I’ve routinely used boost polygon90 with 100k-1m slightly overlapping (routing + via + stdcell pins) 32 bit polygon and found it to be very fast. So I would encourage you to use larger test cases then the ones in your benchmarks to judge performance.

**Reply by NailxSharipov** [reply]
Thanks. I realized that the `N` value in the benchmarks was misleading. It is not the number of polygons, but a parameter used to generate it.

For example, the first benchmark at its maximum `N` contains about **8.4 million polygons**.

I've updated the benchmark page to show how the polygon count is calculated.

**Comment by maliberty** [comment]
What is the difference between iOverlay and xOverlay?  
How many threads in the MULTI runs?  How does the result scale with number of threads?
Did you verify that all tools give the same answer in all cases?
It would be easier to read your tables if you used the same units across a row.
In your tests where number of polygons is linear with N you should use bigger N values.

What is the conceptual advantage of your tool over Boost?

**Reply by NailxSharipov** [reply]
iOverlay is my general-purpose library that works with arbitrary 2D geometry. xOverlay is an attempt to develop a faster algorithm specialized for orthogonal geometry.

iOverlay is a reliable library. It has been battle-tested across many projects and has over 5 million downloads on crates.io.

So far, xOverlay has only been tested by me. However, I have run extensive randomized differential tests comparing its output with iOverlay. The comparison checks the total area, the number of shapes, contours, and points.

The primary advantage of xOverlay is performance. A secondary advantage is access to the planar graph produced by the algorithm.

Sometimes the resulting geometry itself is not what you need. Instead, you may need to know the topological relationships between elements for example, whether two shapes share an edge. This could be useful for connectivity analysis.

**Comment by maliberty** [comment]
Bloat and shrink are also needed operations.

**Reply by NailxSharipov** [reply]
It's not a problem, I've already did it for a general solution.

**Comment by maliberty** [comment]
Do you plan to add support for 45 degree edges?  They do turn up in RDL even in advanced nodes.

**Reply by NailxSharipov** [reply]
Supporting 45 degree geometry is more challenging. It looks like the approach I currently use for orthogonal geometry may not work in this case.
Still, I expect that a specialized implementation could be 2x - 4x faster than the general purpose iOverlay.
Since 45 degree geometry does not appear to be a bottleneck in current designs, I'm not focusing my efforts on it at the moment.
